### PR TITLE
coll/ucc: add support for gather(v), scatterv, reduce_scatter

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
Add support for more collectives in coll ucc: gather, gatherv, reduce_scatter, reduce_scatter_block, scatterv
Signed-off-by: Sergey Lebedev <sergeyle@nvidia.com>

### CODE REVIEW

**Review by bureddy — APPROVED** [review]

**Comment by edgargabriel** [comment]
a small question just for clarification: it seems like this commit adds support for scatterv, but not for scatter, is this in purpose? (the commit does add however support for gather and gatherv)

File: ompi/mca/coll/ucc/coll_ucc_gatherv.c
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_gatherv.c:120]: previous_igather**V**_module

File: ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c:25]: UCC_VERBOSE(5, "fallback ... ") ?

File: ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c:21]: make total_count size_t

File: ompi/mca/coll/ucc/coll_ucc_reduce_scatter_block.c
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_reduce_scatter_block.c:47]: rcount * comm_size -> ((size_t)rcount) * comm_size to avoid potential INT overrun

**Comment by Sergei-Lebedev** [comment]
> a small question just for clarification: it seems like this commit adds support for scatterv, but not for scatter, is this on purpose? (the commit does add e.g. support for gather and gatherv)

yes, in UCC we don't have generic scatter implementation so far. Only device specific TLs (NCCL, RCCL, CUDA) supports it.

**Review by vspetrov — CHANGES_REQUESTED** [review]

**Comment by devreal** [comment]
>> a small question just for clarification: it seems like this commit adds support for scatterv, but not for scatter, is this on purpose? (the commit does add e.g. support for gather and gatherv)
>
> yes, in UCC we don't have generic scatter implementation so far. Only device specific TLs (NCCL, RCCL, CUDA) supports it.

Scatter is a specialization of scatterv. Why not implement the former in terms of the latter?

**Review by vspetrov — APPROVED** [review]

**Comment by janjust** [comment]
We're holding off until scater is added as a commit, then I'll merge

**Comment by Sergei-Lebedev** [comment]
@edgargabriel @devreal added scatter in last commit

**Comment by edgargabriel** [comment]
> @edgargabriel @devreal added scatter in last commit

excellent, thank you very much!

**Pull request merged**
